internal/errors.go

```diff
@@ -0,0 +1,80 @@
+package internal
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"go.temporal.io/sdk/temporal"
+)
+
+// errorType identifies a custom Cadence error as an AWS SDK error.
+const errorType = "AWS_SDK_ERROR"
+
+// EncodedError is a serialized version of the AWS SDK error.
+type EncodedError struct {
+	Code    string
+	Message string
+	OrigErr string
+}
+
+// Error implements the error interface.
+func (e EncodedError) Error() string {
+	return fmt.Sprintf("%s: %s", e.Code, e.Message)
+}
+
+// EncodeError encodes an error returned by the AWS SDK as a Cadence error.
+func EncodeError(err error) error {
+	if err == nil {
+		return nil
+	}
+
+	if aerr, ok := err.(awserr.Error); ok {
+		newErr := EncodedError{
+			Code:    aerr.Code(),
+			Message: aerr.Message(),
+		}
+
+		// TODO: try to encode the original error as well?
+		if aerr.OrigErr() != nil {
+			newErr.OrigErr = aerr.OrigErr().Error()
+		}
+
+		return temporal.NewApplicationError(newErr.Error(), errorType, newErr)
+	}
+
+	return err
+}
+
+// DecodedError is a serialized version of the AWS SDK error.
+type DecodedError struct {
```

Comment:
I don't like the name. On the wire, it will be encodeded decoded error :).
What about AwsError?

Reply:
Doesn't matter that much, users would not interact with it directly anyway. (This is just an intermediary struct) Actually, it shouldn't even be exported. Nor should the encoded error.